src/FourMomenta.cxx

```diff
@@ -224,29 +224,35 @@ std::vector<FourVector<double> > calculate_four_momenta(double initial_mass, con
     double m2_sum_1 = 0;
     double m_sum_1 = initial_mass;
     for (size_t i = 0; i < n_fsp; ++i) {
-        pp[i][i] = pow(FSPs[i]->mass(), 2);
+        const auto m = FSPs[i]->mass();
+
+        pp[i][i] = m * m;
         m2_sum_1 += pp[i][i];
-        m_sum_1 -= FSPs[i]->mass();
+        m_sum_1 -= m;
     }
 
     // add two-particle invariant masses for those provided
     double m2_sum_2 = 0;
     for (size_t i = 0; i < axes.size(); ++i) {
-        if (axes[i]->indices()[0] < axes[i]->indices()[1])
-            pp[axes[i]->indices()[0]][axes[i]->indices()[1]] = squared_masses[i];
-        else
-            pp[axes[i]->indices()[1]][axes[i]->indices()[0]] = squared_masses[i];
+        const auto row = std::min(axes[i]->indices()[0], axes[i]->indices()[1]);
+        const auto col = std::max(axes[i]->indices()[0], axes[i]->indices()[1]);
+        pp[row][col] = squared_masses[i];
+
         m2_sum_2 += squared_masses[i];
     }
 
     //////////////////////////////////////////////////
     // find unset two-particle invariant mass
     // and calculate fsp squared masses
-    for (unsigned i = 0; i < n_fsp; ++i)
-        for (unsigned j = i + 1; j < n_fsp; ++j)
-            if (pp[i][j] < 0)
-                // calculate unset mass: m^2_missing = M^2_isp + (n_fsp - 2) * sum_fsp (m^2) - sum_given(m^2)
-                pp[i][j] = pow(initial_mass, 2) + (n_fsp - 2.) * m2_sum_1 - m2_sum_2;
+    for (unsigned i = 0; i < n_fsp; ++i) {
+        // Find element that are smaller than zero in the upper triangle of the matrix.
+        const auto it = std::find_if(std::begin(pp[i]) + i + 1, std::end(pp[i]),
```

Comment:
This is actually ___wrong.___ It will only assign to the first negative element for each value of _i_. Still, all the tests pass.